Wasm compilation error

Some errors have been commited despite the possibility to check before hand.

The following errors and maybe mire are seen:
﻿2025-07-20T20:37:18.6983650Z ##[group]Run ./build-wasm.sh
./build-wasm.sh
shell: /usr/bin/bash -e {0}
env:
  GITHUB_PAGES: true
🚀 Building WASM Task Manager...
📦 Building WASM package...
[INFO]: 🎯  Checking for the Wasm target...
info: downloading component 'rust-std' for 'wasm32-unknown-unknown'
info: installing component 'rust-std' for 'wasm32-unknown-unknown'
[INFO]: 🌀  Compiling to Wasm...
   Compiling autocfg v1.5.0
   Compiling once_cell v1.21.3
   Compiling version_check v0.9.5
   Compiling bytemuck_derive v1.10.0
   Compiling wasm-bindgen v0.2.100
   Compiling zerocopy v0.8.26
   Compiling lock_api v0.4.13
   Compiling parking_lot_core v0.9.11
   Compiling js-sys v0.3.77
   Compiling ahash v0.8.12
   Compiling bytemuck v1.23.1
   Compiling scopeguard v1.2.0
   Compiling ttf-parser v0.25.1
   Compiling smallvec v1.15.1
   Compiling emath v0.30.0
   Compiling ab_glyph_rasterizer v0.1.9
   Compiling parking_lot v0.12.4
   Compiling ecolor v0.30.0
   Compiling slotmap v1.0.7
   Compiling nohash-hasher v0.2.0
   Compiling log v0.4.27
   Compiling owned_ttf_parser v0.25.0
   Compiling ab_glyph v0.2.31
   Compiling epaint_default_fonts v0.30.0
   Compiling profiling v1.0.17
   Compiling memoffset v0.9.1
   Compiling epaint v0.30.0
   Compiling egui v0.30.0
   Compiling web-sys v0.3.77
   Compiling num-traits v0.2.19
   Compiling litrs v0.4.1
   Compiling document-features v0.2.11
   Compiling web-time v1.1.0
   Compiling wasm-bindgen-futures v0.4.50
   Compiling raw-window-handle v0.6.2
   Compiling percent-encoding v2.3.1
   Compiling static_assertions v1.1.0
   Compiling egui_plot v0.30.0
   Compiling chrono v0.4.41
   Compiling glow v0.16.0
   Compiling egui_glow v0.30.0
   Compiling eframe v0.30.0
   Compiling gh_actions v0.1.0 (/home/runner/work/gh_actions/gh_actions)
error: unused imports: `DateTime`, `Duration`, and `Utc`
 --> src/kpi_app.rs:3:14
  |
3 | use chrono::{DateTime, Utc, Duration};
  |              ^^^^^^^^  ^^^  ^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

error: unused import: `crate::task::TaskManager`
 --> src/kpi_app.rs:4:5
  |
4 | use crate::task::TaskManager;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^

error: unused import: `std::collections::HashMap`
 --> src/kpi_app.rs:5:5
  |
5 | use std::collections::HashMap;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/wasm.rs:178:32
    |
178 |                 Box::new(|_cc| Box::new(TaskManagerApp::new())),
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Result<Box<dyn App>, Box<...>>`, found `Box<TaskManagerApp>`
    |
    = note: expected enum `Result<Box<(dyn App + 'static)>, Box<(dyn StdError + Send + Sync + 'static)>>`
             found struct `Box<TaskManagerApp>`
help: try wrapping the expression in `Ok`
    |
178 |                 Box::new(|_cc| Ok(Box::new(TaskManagerApp::new()))),
    |                                +++                               +

error[E0308]: mismatched types
   --> src/wasm.rs:176:17
    |
175 |             .start(
    |              ----- arguments to this method are incorrect
176 |                 canvas_id,
    |                 ^^^^^^^^^ expected `HtmlCanvasElement`, found `&str`
    |
note: method defined here
   --> /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/eframe-0.30.0/src/web/web_runner.rs:53:18
    |
53  |     pub async fn start(
    |                  ^^^^^

error: unused import: `egui`
 --> src/app.rs:1:5
  |
1 | use egui::*;
  |     ^^^^

error: unused variable: `frame`
  --> src/app.rs:69:59
   |
69 |     fn show_kpi_dashboard(&mut self, ctx: &egui::Context, frame: &mut eframe::Frame) {
   |                                                           ^^^^^ help: if this is intentional, prefix it with an underscore: `_frame`
   |
   = note: `-D unused-variables` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(unused_variables)]`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `gh_actions` (lib) due to 7 previous errors
Error: Compiling your crate to WebAssembly failed
Caused by: Compiling your crate to WebAssembly failed
Caused by: failed to execute `cargo build`: exited with exit status: 101
  full command: cd "/home/runner/work/gh_actions/gh_actions" && "cargo" "build" "--lib" "--release" "--target" "wasm32-unknown-unknown"

--- a/src/app.rs
+++ b/src/app.rs
@@ -1,4 +1,6 @@
-use egui::*;
+use egui::{Context, CentralPanel, Layout, Align, ScrollArea, Color32};
+use eframe::App;
+use egui_plot::{Line, Plot, PlotPoints};
 use crate::kpi_app::KpiApp;
 
 #[derive(Default)]
@@ -23,8 +25,8 @@ impl TaskManagerApp {
     }
 }
 
-impl eframe::App for TaskManagerApp {
-    fn update(&mut self, ctx: &egui::Context, frame: &mut eframe::Frame) {
+impl App for TaskManagerApp {
+    fn update(&mut self, ctx: &Context, frame: &mut eframe::Frame) {
         match self.current_view {
             AppView::TaskManager => {
                 self.show_task_manager(ctx, frame);
@@ -37,11 +39,11 @@ impl eframe::App for TaskManagerApp {
 }
 
 impl TaskManagerApp {
-    fn show_task_manager(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
-        egui::CentralPanel::default().show(ctx, |ui| {
+    fn show_task_manager(&mut self, ctx: &Context, _frame: &mut eframe::Frame) {
+        CentralPanel::default().show(ctx, |ui| {
             ui.horizontal(|ui| {
                 ui.heading("🚀 WASM Task Manager");
-                ui.with_layout(egui::Layout::right_to_left(egui::Align::Center), |ui| {
+                ui.with_layout(Layout::right_to_left(Align::Center), |ui| {
                     if ui.button("📊 View KPIs").clicked() {
                         self.current_view = AppView::KpiDashboard;
                     }
@@ -66,11 +68,11 @@ impl TaskManagerApp {
         });
     }
     
-    fn show_kpi_dashboard(&mut self, ctx: &egui::Context, frame: &mut eframe::Frame) {
-        egui::CentralPanel::default().show(ctx, |ui| {
+    fn show_kpi_dashboard(&mut self, ctx: &Context, _frame: &mut eframe::Frame) {
+        CentralPanel::default().show(ctx, |ui| {
             ui.horizontal(|ui| {
                 ui.heading("📊 Task Management KPIs");
-                ui.with_layout(egui::Layout::right_to_left(egui::Align::Center), |ui| {
+                ui.with_layout(Layout::right_to_left(Align::Center), |ui| {
                     if ui.button("← Back to Tasks").clicked() {
                         self.current_view = AppView::TaskManager;
                     }
@@ -167,7 +169,7 @@ impl TaskManagerApp {
                 if tasks.is_empty() {
                     ui.label("No tasks yet. Add one above!");
                 } else {
-                    egui::ScrollArea::vertical().show(ui, |ui| {
+                    ScrollArea::vertical().show(ui, |ui| {
                         for task in tasks.iter() {
                             ui.group(|ui| {
                                 ui.horizontal(|ui| {
@@ -182,7 +184,7 @@ impl TaskManagerApp {
                                         }
                                     });
                                     
-                                    ui.with_layout(egui::Layout::right_to_left(egui::Align::Center), |ui| {
+                                    ui.with_layout(Layout::right_to_left(Align::Center), |ui| {
                                         if ui.button("🗑").clicked() {
                                             crate::wasm::remove_task(task.id);
                                         }
@@ -276,7 +278,7 @@ impl TaskManagerApp {
             .show(ui, |plot_ui| {
                 plot_ui.line(
                     Line::new(PlotPoints::from(points))
-                        .color(egui::Color32::from_rgb(100, 200, 100))
+                        .color(Color32::from_rgb(100, 200, 100))
                         .name("Tasks Completed")
                 );
             });

--- a/src/kpi_app.rs
+++ b/src/kpi_app.rs
@@ -1,8 +1,5 @@
 use egui::*;
 use egui_plot::{Line, Plot, PlotPoints};
-use chrono::{DateTime, Utc, Duration};
-use crate::task::TaskManager;
-use std::collections::HashMap;
 
 #[derive(Default)]
 pub struct KpiApp {

--- a/src/wasm.rs
+++ b/src/wasm.rs
@@ -1,7 +1,9 @@
 use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
 use crate::task::{TaskManager, Task};
 use crate::app::TaskManagerApp;
 use std::sync::Mutex;
+use web_sys::HtmlCanvasElement;
 
 // Global task manager instance
 lazy_static::lazy_static! {
@@ -170,12 +172,20 @@ pub fn start_egui_app(canvas_id: &str) {
     
     let web_options = eframe::WebOptions::default();
     
-    wasm_bindgen_futures::spawn_local(async {
+    wasm_bindgen_futures::spawn_local(async move {
+        let window = web_sys::window().unwrap();
+        let document = window.document().unwrap();
+        let canvas = document
+            .get_element_by_id(canvas_id)
+            .unwrap()
+            .dyn_into::<HtmlCanvasElement>()
+            .unwrap();
+        
         let result = eframe::WebRunner::new()
             .start(
-                canvas_id,
+                canvas,
                 web_options,
-                Box::new(|_cc| Box::new(TaskManagerApp::new())),
+                Box::new(|_cc| Ok(Box::new(TaskManagerApp::new()))),
             )
             .await;
             

